unit_tests/test_redis_aggregator.py

```diff
@@ -0,0 +1,238 @@
+# pylint: disable=missing-function-docstring, protected-access
+import json
+import uuid
+from unittest.mock import patch, PropertyMock, MagicMock
+import pytest
+
+from gsy_e_sdk.redis_aggregator import RedisAggregator, RedisAggregatorAPIException
+
+TEST_AGGREGATOR_NAME = "TestAgg"
+
+TEST_AREA_NAME = "TestArea"
+TEST_AREA_MAPPING = {TEST_AREA_NAME: [str(uuid.uuid4())]}
+
+TEST_TRANSACTION_ID = str(uuid.uuid4())
+TEST_AGGREGATOR_UUID = str(uuid.uuid4())
+
+TEST_DEVICE_UUID_1 = str(uuid.uuid4())
+TEST_DEVICE_UUID_2 = str(uuid.uuid4())
+TEST_BATCH_COMMAND = "TEST_BATCH_COMMAND"
+
+TEST_BATCH_COMMAND_DICT = {
+    TEST_DEVICE_UUID_1: [TEST_BATCH_COMMAND],
+    TEST_DEVICE_UUID_2: [TEST_BATCH_COMMAND]
+}
+
+TEST_TRANSACTION_ID_RESPONSE = "TEST_TRANSACTION_ID_RESPONSE"
+TEST_TRANSACTION_ID_BUFFER = {
+    TEST_TRANSACTION_ID: {"transaction_id": TEST_TRANSACTION_ID,
+                          "response": TEST_TRANSACTION_ID_RESPONSE}
+}
+
+TEST_RESPONSE = "TEST_RESPONSE"
+TEST_TRANSACTION_ID_RESPONSE_BUFFER = {TEST_TRANSACTION_ID: TEST_RESPONSE}
+
+
+@pytest.fixture(autouse=True)
+def fixture_mock_connections(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.StrictRedis")
+    mocker.patch("gsy_e_sdk.redis_aggregator.ThreadPoolExecutor")
+
+
+@pytest.fixture(name="mock_grid_fee_calculation")
+def fixture_mock_grid_fee_calculation(mocker):
+    """Mock GridFeeCalculation class."""
+    mocker.patch("gsy_e_sdk.redis_aggregator.GridFeeCalculation")
+
+
+@pytest.fixture(name="mock_transaction_id_and_timeout_blocking")
+def fixture_mock_transaction_id_and_timeout_blocking(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.uuid.uuid4",
+                 return_value=TEST_TRANSACTION_ID)
+    mocker.patch("gsy_e_sdk.redis_aggregator.wait_until_timeout_blocking")
+
+
+@pytest.fixture(name="mock_client_command_buffer_attributes")
+def fixture_mock_client_command_buffer(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.ClientCommandBuffer.buffer_length",
+                 new_callable=PropertyMock, return_value=3)
+    mocker.patch("gsy_e_sdk.redis_aggregator.ClientCommandBuffer.execute_batch",
+                 return_value=TEST_BATCH_COMMAND_DICT)
+
+
+@pytest.fixture(name="aggregator")
+def fixture_aggregator(mocker):
+    mocker.patch("gsy_e_sdk.redis_aggregator.RedisAggregator._create_aggregator")
+    return RedisAggregator(aggregator_name=TEST_AGGREGATOR_NAME)
+
+
+class TestRedisAggregatorConstructor:
+    """Includes tests for the constructor of the RedisAggregator class."""
+
+    @staticmethod
+    @pytest.mark.usefixtures("mock_grid_fee_calculation")
+    def test_constructor_grid_fee_calculation_instantiated(aggregator):
+        assert aggregator.grid_fee_calculation is not None
+
+    @staticmethod
+    def test_constructor_redis_db_instantiated(aggregator):
+        assert aggregator.redis_db is not None
```

Comment:
If we really want to test constructors, we should probably check the details of what's happening (not only checking for `is None`). 

Here, for example, we want to make sure that `StrictRedis.from_url` is called with the `redis_url` parameter, and that the result of this operation is stored inside of `self.redis_db`.

Reply:
The same concept applies to `test_constructor_grid_fee_calculation_instantiated` above.

Reply:
To address your comment, I have added one more assertion about the object being called. 
However, I'm not sure on how to assert that a certain attribute (set in the constructor) stores the mocked object in this case. 
I have tried the following for example:
```
    def test_constructor_redis_db_instantiated():
        with patch("gsy_e_sdk.redis_aggregator.StrictRedis"
                   ) as mocked_class:
            agg = RedisAggregator(aggregator_name=TEST_AGGREGATOR_NAME)
            mocked_class.from_url.assert_called_with(LOCAL_REDIS_URL)
            assert agg.redis_db is mocked_class
```

This kind of test fails, because AFAIS, when the aggregator is created, a _StrictRedis-mocked_ object is instantiated. When I look for the assertion to see wether the attribute has this mock stored in there, another mock object is created and used to compare. As these two are different, then the test fails. 

Do you have any suggestion on how I could assert that a mocked object is stored in a class attribute?

Reply:
What we're trying to do here is basically stubbing the return value of `StrictRedis.from_url()`. So we don't need to mock the entire `StrictRedis` object. I took the liberty to try this out, and this seems to work:
```python
@staticmethod
def test_constructor_redis_db_instantiated(aggregator):
    redis_db_mock = MagicMock()
    with patch("gsy_e_sdk.redis_aggregator.StrictRedis.from_url", return_value=redis_db_mock
           ) as from_url_mock:
        agg = RedisAggregator(aggregator_name=TEST_AGGREGATOR_NAME, redis_url="some-url")
        from_url_mock.assert_called_with("some-url")
        assert agg.redis_db is redis_db_mock
```
As you can see, I've instantiated the aggregator passing the `redis_url` argument, because that value will be used to call `StrictRedis.from_url()`. 
The other thing I've done is explicitly creating a `MagicMock` instance, so that we could re-use it in the end to make sure that it's the same instance used to populate the `self.redis_db` variable.

Let me know if this makes sense or if you'd rather change something.
P.S.: Pylint complains about the `aggregator` parameter but I didn't change it in the example to save some time.

Reply:
Ok, that makes sense to me. Will replicate that in the other tests. Thanks!